Port processing vector.features to c++

nyalldawson opened the pull request:
(Relies on #4402)

This PR removes processing vector.features, having first completely ported this to the c++ methods QgsProcessingUtils.getFeatures and QgsProcessingUtils.featureCount

Also includes restoration of the exception upon encountering invalid geometries (if processing config is set to abort on invalid geometries)

nyalldawson committed:
Create a context object for processing algorithm execution

nyalldawson committed:
[processing] Pass context argument to processAlgorithm

nyalldawson committed:
Add UseSelection flag to QgsProcessingContext

Indicates whether only selected features should be used in
algorithms

nyalldawson committed:
Add invalid geometry handling method to processing context

nyalldawson committed:
Add context argument to vector.features

nyalldawson committed:
Add context argument to other vector methods

nyalldawson committed:
Use context to control vector.features()

nyalldawson committed:
Add methods for getting vector iterators to QgsProcessingUtils

nyalldawson committed:
Allow setting the invalid geometry callback in processing context

nyalldawson committed:
Raise GeoAlgorithmExecutionException when encountering invalid
geometries in QgsProcessingUtils::getFeatures

nyalldawson committed:
Remove len method from vector.features

Use QgsProcessingUtils::featureCount instead

nyalldawson committed:
Remove vector.features

Use QgsProcessingUtils.getFeatures instead

m-kuhn commented on src/core/processing/qgsprocessingutils.cpp:
```diff
@@ -172,3 +173,29 @@ QString QgsProcessingUtils::normalizeLayerSource( const QString &source )
+
+long QgsProcessingUtils::featureCount( QgsVectorLayer *layer, const QgsProcessingContext &context )
+{
+  bool useSelection = context.flags() & QgsProcessingContext::UseSelection && layer->selectedFeatureCount() > 0;
```
The `layer->selectedFeatureCount() > 0` condition might lead to surprising results in algorithms that set a selection and then continue to process based on this selection. If a select by rectangle doesn't select anything, this will result in processing all features instead of none.

m-kuhn replied:
This should preferably be done by applying a filter anyway - or the algorithm shouldn't be executed in a background task if it modifies the layer state.

nyalldawson replied:
I agree, but it's the behavior with current processing and I don't think users would appreciate the removal of this feature.

m-kuhn replied:
Then this will need to be added as `QgsProcessingContext::UseSelectionIfPresent` to be explicit. I really fear that this will cause data corruption.

nirvn replied:
I would appreciate getting rid of the current behavior and replace it with a superior one :) -- I find that a processing-wide single [x] use only selected flag is really bad. Beyond offering absolutely no visible feedback, it's not practical. 

I'd rather have some sort of a checkbox attached to the layer selection widget, which would offer a use-selected-only UI at hand.

nyalldawson replied:
Could we leave the discussion about this "feature" out of this PR? Because:

- it's the current behaviour, and has been for a while, so changing it should first be discussed on the dev list
- I don't want this PR held up with that discussion - I'd rather get this merged and then we can refine/change based on the other discussion

For now this is a direct copy of the existing logic

nirvn replied:
@nyalldawson , yeah, no objection to the PR per say.

nyalldawson committed:
Add missing doc

m-kuhn commented on python/plugins/processing/tools/vector.py:
```diff
@@ -354,19 +295,15 @@ def simpleMeasure(geom, method=0, ellips=None, crs=None):
 
 
-def getUniqueValues(layer, fieldIndex):
+def getUniqueValues(layer, context, fieldIndex):
```
Should that be using `QgsVectorLayer::uniqueValues` directly?

nyalldawson replied:
Baby steps... :)

nyalldawson replied:
Done

nyalldawson committed:
Rename UseSelection -> UseSelectionIfPresent, clarify docs

nyalldawson committed:
Move createContext to dataobjects

nyalldawson committed:
Fix failing tests

nyalldawson committed:
Indentation/sipify

nyalldawson committed:
[processing] Don't use config values for invalid geometry handling in tests

nyalldawson committed:
Add c++ optimised uniqueValues method which respects processing context

Remove processing vector.uniqueValues/vector.getUniqueValues
and port usage to c++ method

Should be much faster than the python method, as the c++ method takes
advantage of handing off the unique values calculation to the
provider source whenever possible

nyalldawson committed:
Fix failing test

nyalldawson committed:
Port some more code to new API

nyalldawson commented:
After offline discussion with @m-kuhn we agreed to merge this as a step on the way forward, and hopefully we can add the per-input layer filters at a later stage prior to 3.0 (when the parameter refactoring is done)
